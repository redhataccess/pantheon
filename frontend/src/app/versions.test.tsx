import React from 'react';
import { Versions } from '@app/versions';
import "isomorphic-fetch"

import { mount, shallow } from 'enzyme';
import { Button, Card, DataList, DataListItem, DataListItemCells, DataListItemRow, DataListToggle, Dropdown, Form, FormGroup, FormSelect, FormSelectOption, InputGroup, InputGroupText, Modal, TextInput, DropdownItem, Title, Alert, AlertActionCloseButton, DataListContent, KebabToggle } from '@patternfly/react-core';
import renderer from 'react-test-renderer';
import sinon from 'sinon'

const anymatch = require('anymatch');

const props = {
    updateDate: (draftUpdateDate,releaseUpdateDate,releaseVersion) => anymatch,
    modulePath: "/modules/test",
    onGetProduct: (productValue) => anymatch,
    onGetVersion: (versionValue) => anymatch,
    versionModulePath: "/modules/test/en_US/1",
}

describe('Versions tests', () => {
    test('should render Versions component', () => {
        const view = shallow(<Versions {...props} />);
        expect(view).toMatchSnapshot();
    });

    it('should render a Button', () => {
        const wrapper = mount(<Versions {...props} />);
        const button = wrapper.find(Button);
        expect(button.exists()).toBe(true)
    });

    it('should render a Card', () => {
        const wrapper = mount(<Versions {...props} />);
        const card = wrapper.find(Card);
        expect(card.exists()).toBe(true)
    });

    it('should render a Data List', () => {
        const wrapper = mount(<Versions {...props} />);
        const dataList = wrapper.find(DataList);
        expect(dataList.exists()).toBe(true)
    });

    it('should render a DataListItem', () => {
        const wrapper = mount(<Versions {...props} />);
        const dataListItem = wrapper.find(DataListItem);
        expect(dataListItem.exists()).toBe(true)
    });

    it('should render a DataListItemCells Element', () => {
        const wrapper = mount(<Versions {...props} />);
        const dataListItemCells = wrapper.find(DataListItemCells);
        expect(dataListItemCells.exists()).toBe(true)
    });

    it('should render a DataListItemRow element', () => {
        const wrapper = mount(<Versions {...props} />);
        const dataListItemRow = wrapper.find(DataListItemRow);
        expect(dataListItemRow.exists()).toBe(true)
    });

    it('should render a DataListToggle', () => {
        const wrapper = mount(<Versions {...props} />);
        const dataListToggle = wrapper.find(DataListToggle);
        expect(dataListToggle.exists()).toBe(true)
    });

    it('should render a DataListContent', () => {
        const wrapper = mount(<Versions {...props} />);
        const dataListContent = wrapper.find(DataListContent);
        expect(dataListContent.exists()).toBe(true)
    });

    it('should render a Modal', () => {
        const wrapper = mount(<Versions {...props} />);
        const modal = wrapper.find(Modal);
        expect(modal.exists()).toBe(true)
    });

    it('should render a Form', () => {
        const wrapper = shallow(<Versions {...props} />);
        wrapper.setState({ 'login': true })
        wrapper.setState({ 'isModalOpen': true })
        const form = wrapper.find(Form);
        expect(form.exists()).toBe(true)
    });

    it('should render a FormGroup', () => {
        const wrapper = shallow(<Versions {...props} />);
        wrapper.setState({ 'login': true })
        wrapper.setState({ 'isModalOpen': true })
        const formGroup = wrapper.find(FormGroup);
        expect(formGroup.exists()).toBe(true)
    });

    it('should render a FormSelect', () => {
        const wrapper = shallow(<Versions {...props} />);
        wrapper.setState({ 'login': true })
        wrapper.setState({ 'isModalOpen': true })
        const formSelect = wrapper.find(FormSelect);
        expect(formSelect.exists()).toBe(true)
    });

    it('should render a FormSelectOption', () => {
        const wrapper = shallow(<Versions {...props} />);
        wrapper.setState({ 'login': true })
        wrapper.setState({ 'isModalOpen': true })
        const formSelectOption = wrapper.find(FormSelectOption);
        expect(formSelectOption.exists()).toBe(true)
    });

    it('should render a InputGroup', () => {
        const wrapper = shallow(<Versions {...props} />);
        wrapper.setState({ 'login': true })
        wrapper.setState({ 'isModalOpen': true })
        const inputGroup = wrapper.find(InputGroup);
        expect(inputGroup.exists()).toBe(true)
    });

    it('should render a success Alert', () => {
        const wrapper = shallow(<Versions {...props} />);
        wrapper.setState({ 'login': true })
        wrapper.setState({ 'successAlertVisble': true })
        const alert = wrapper.find(Alert);
        expect(alert.exists()).toBe(true)
    });

    it('should render a Button', () => {
        const wrapper = mount(<Versions {...props} />);
        const button = wrapper.find(Button);
        expect(button.exists()).toBe(true)
    });

    it('should render a KebabToggle', () => {
        const wrapper = mount(<KebabToggle />);
    });

    it('test fetchVersions function', () => {
        const wrapper = renderer.create(<Versions {...props} />);
        const inst = wrapper.getInstance();
        expect(inst.fetchVersions()).toMatchSnapshot();
    });

    it('test changePublishState function', () => {
        const wrapper = renderer.create(<Versions {...props} />);
        const inst = wrapper.getInstance();
        expect(inst.changePublishState("Publish")).toMatchSnapshot();
    });

    it('test onArchiveSelect function', () => {
        const wrapper = renderer.create(<Versions {...props} />);
        const inst = wrapper.getInstance();
        expect(inst.onArchiveSelect()).toMatchSnapshot();
    });

    it('test onHeadingToggle function', () => {
        const wrapper = renderer.create(<Versions {...props} />);
        const inst = wrapper.getInstance();
        expect(inst.onHeadingToggle()).toMatchSnapshot();
    });

    it('test onExpandableToggle function', () => {
        const wrapper = renderer.create(<Versions {...props} />);
        const inst = wrapper.getInstance();
        const data = [{"isDropdownOpen": true}]
        expect(inst.onExpandableToggle(data)).toMatchSnapshot();
    });

    it('test onArchiveToggle function', () => {
        const wrapper = renderer.create(<Versions {...props} />);
        const inst = wrapper.getInstance();
        const data = [{"isDropdownOpen": true}]
        expect(inst.onArchiveToggle(data)).toMatchSnapshot();
    });

    it('test previewDoc function', () => {
        const wrapper = renderer.create(<Versions {...props} />);
        const inst = wrapper.getInstance();
        expect(inst.previewDoc()).toMatchSnapshot();
    });

    it('test saveMetadata function', () => {
        const wrapper = renderer.create(<Versions {...props} />);
        const inst = wrapper.getInstance();
        expect(inst.saveMetadata()).toMatchSnapshot();
    });

    it('test onChangeProduct function', () => {
        const wrapper = renderer.create(<Versions {...props} />);
        const inst = wrapper.getInstance();
        expect(inst.onChangeProduct()).toMatchSnapshot();
    });

    it('test onChangeUsecase function', () => {
        const wrapper = renderer.create(<Versions {...props} />);
        const inst = wrapper.getInstance();
        expect(inst.onChangeUsecase()).toMatchSnapshot();
    });

    it('test handleURLInput function', () => {
        const wrapper = renderer.create(<Versions {...props} />);
        const inst = wrapper.getInstance();
        expect(inst.handleURLInput()).toMatchSnapshot();
    });

    it('test getModuleUrl function', () => {
        const wrapper = renderer.create(<Versions {...props} />);
        const inst = wrapper.getInstance();
        expect(inst.getModuleUrl()).toMatchSnapshot();
    });

    it('test renderRedirect function', () => {
        const wrapper = renderer.create(<Versions {...props} />);
        const inst = wrapper.getInstance();
        expect(inst.renderRedirect()).toMatchSnapshot();
    });

    it('test loginRedirect function', () => {
        const wrapper = renderer.create(<Versions {...props} />);
        const inst = wrapper.getInstance();
        expect(inst.loginRedirect()).toMatchSnapshot();
    });

    it('test dismissNotification function', () => {
        const wrapper = renderer.create(<Versions {...props} />);
        const inst = wrapper.getInstance();
        expect(inst.dismissNotification()).toMatchSnapshot();
    });

    it('test getMetadata function', () => {
        const wrapper = shallow(<Versions {...props} />);
        const instance = wrapper.instance();
        wrapper.setState({ 'login': true })
        wrapper.setState({ 'isModalOpen': true })
        // Assuming metadata exists
        const spy = sinon.spy(instance, 'saveMetadata');
        const urlFragment = wrapper.find('input');
        expect(urlFragment.exists()).toBe(true)

        const useCaseValue = wrapper.find('[aria-label="FormSelect Usecase"]').simulate('change', { target: { value: 'Administer' } });
        expect(useCaseValue.exists()).toBe(true)

        wrapper.setState({ 'moduleUrl': urlFragment })
        expect(wrapper.state('moduleUrl')).toBeDefined()

        wrapper.setState({ 'usecaseValue': useCaseValue })
        expect(instance.state['usecases'][0]).toEqual('Administer')
    });

    it('test getHarrayChildNamed function', () => {
        const wrapper = renderer.create(<Versions {...props} />);
        const inst = wrapper.getInstance();
        expect(inst.getHarrayChildNamed("__children__")).toMatchSnapshot();
    });

    it('test fetchProductVersionDetails function', () => {
        const wrapper = renderer.create(<Versions {...props} />);
        const inst = wrapper.getInstance();
        expect(inst.fetchProductVersionDetails()).toMatchSnapshot();
    });

    it('test hideSuccessAlert function', () => {
        const wrapper = renderer.create(<Versions {...props} />);
        const inst = wrapper.getInstance();
        expect(inst.hideSuccessAlert()).toMatchSnapshot();
    });

    test('Version Button click', () => {
        const wrapper = shallow(<Versions {...props} />);
        const instance = wrapper.instance();
        const spy = sinon.spy(instance, 'onHeadingToggle');

        wrapper.find(DataListToggle).simulate('click');
        sinon.assert.called(spy);
    });

    it('should handle state changes for changePublishState', () => {
        const wrapper = shallow(<Versions {...props} />)

        expect(wrapper.state('changePublishState')).toBe(false)
        wrapper.setState({ 'changePublishState': true })
        expect(wrapper.state('changePublishState')).toBe(true)
    });


    it('should handle state changes for login', () => {
        const wrapper = shallow(<Versions {...props} />)

        expect(wrapper.state('login')).toBe(false)
        wrapper.setState({ 'login': true })
        expect(wrapper.state('login')).toBe(true)
    });

    it('should handle state changes for initialLoad', () => {
        const wrapper = shallow(<Versions {...props} />)

        expect(wrapper.state('initialLoad')).toBe(true)
        wrapper.setState({ 'initialLoad': false })
        expect(wrapper.state('initialLoad')).toBe(false)
    });

    it('should handle state changes for versionSelected', () => {
        const wrapper = shallow(<Versions {...props} />)

        expect(wrapper.state('versionSelected')).toBe('')
        wrapper.setState({ 'versionSelected': "Please select a version" })
        expect(wrapper.state('versionSelected')).toBe("Please select a version")
    });

    it('should handle state changes for versionUUID', () => {
        const wrapper = shallow(<Versions {...props} />)

        expect(wrapper.state('versionUUID')).toBe("")
        wrapper.setState({ 'versionUUID': "Select a Version" })
        expect(wrapper.state('versionUUID')).toBe("Select a Version")
    });

    it('should handle state changes for versionValue', () => {
        const wrapper = shallow(<Versions {...props} />)

        expect(wrapper.state('versionValue')).toBe('')
        wrapper.setState({ 'versionValue': "version value" })
        expect(wrapper.state('versionValue')).toBe("version value")
    });

    it('should handle state changes for isHeadingToggle', () => {
        const wrapper = shallow(<Versions {...props} />)

        expect(wrapper.state('isHeadingToggle')).toBe(true)
        wrapper.setState({ 'isHeadingToggle': false })
        expect(wrapper.state('isHeadingToggle')).toBe(false)
    });

    // Value testing with Enzyme.
    it('renders Version heading', () => {
        const wrapper = mount(<Versions {...props} />);
        const sourceTypeText = wrapper.find('#span-source-type-version').text();

        // ensure it matches what is expected
        expect(sourceTypeText).toEqual("Version");
    });

    it('renders Published heading', () => {
        const wrapper = mount(<Versions {...props} />);
        const sourceTypeText = wrapper.find('#span-source-type-version-published').text();

        // ensure it matches what is expected
        expect(sourceTypeText).toEqual("Published");
    });

    it('renders Draft Uploaded heading', () => {
        const wrapper = mount(<Versions {...props} />);
        const sourceTypeText = wrapper.find('#span-source-type-version-draft-uploaded').text();

        // ensure it matches what is expected
        expect(sourceTypeText).toEqual("Draft Uploaded");
    });

    it('has a props', () => {
        const versions = mount(<Versions {...props} />).matchesElement
        expect(versions.length === 1)
    });

    it('test getHarrayChildNamed function', () => {
        const wrapper = renderer.create(<Versions {...props} />);
        const inst = wrapper.getInstance();
        expect(inst.getHarrayChildNamed(anymatch, 'metadata')).toMatchSnapshot();
    });
});