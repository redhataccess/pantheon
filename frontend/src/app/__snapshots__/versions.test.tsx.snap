// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Versions tests should render Versions component 1`] = `
<Fragment>
  <Card>
    <div>
      <DataList
        aria-label="Simple data list"
      >
        <DataListItem
          aria-labelledby="simple-item1"
          isExpanded={true}
        >
          <DataListItemRow
            id="data-rows-header"
          >
            <DataListToggle
              aria-controls="width-ex3-expand1"
              id="width-ex3-toggle1"
              isExpanded={true}
              onClick={[Function]}
            />
            <DataListItemCells
              dataListCells={
                Array [
                  <DataListCell>
                    <span
                      className="sp-prop-nosort"
                      id="span-source-type-version"
                    >
                      Version
                    </span>
                  </DataListCell>,
                  <DataListCell>
                    <span
                      className="sp-prop-nosort"
                      id="span-source-type-version-published"
                    >
                      Published
                    </span>
                  </DataListCell>,
                  <DataListCell>
                    <span
                      className="sp-prop-nosort"
                      id="span-source-type-version-draft-uploaded"
                    >
                      Draft Uploaded
                    </span>
                  </DataListCell>,
                  <DataListCell>
                    <span
                      className="sp-prop-nosort"
                      id="span-source-name-version-publish-buttons"
                    />
                  </DataListCell>,
                  <DataListCell>
                    <span
                      className="sp-prop-nosort"
                      id="span-source-name"
                    />
                  </DataListCell>,
                ]
              }
            />
          </DataListItemRow>
        </DataListItem>
        <DataListContent
          aria-label="Secondary Content Details"
          id="Content"
          isHidden={false}
          noPadding={true}
        />
      </DataList>
    </div>
  </Card>
  <Modal
    actions={
      Array [
        <Button
          form="edit_metadata"
          onClick={[Function]}
          variant="primary"
        >
          Save
        </Button>,
        <Button
          onClick={[Function]}
          variant="secondary"
        >
          Cancel
        </Button>,
      ]
    }
    ariaDescribedById="edit-metadata"
    className=""
    header={
      <React.Fragment>
        <Title
          headingLevel="h1"
          size="2xl"
        >
          Edit Metadata
        </Title>
        <br />
        <p
          className="pf-u-pl-sm"
        >
          All fields are required.
        </p>
      </React.Fragment>
    }
    hideTitle={false}
    isLarge={false}
    isOpen={false}
    isSmall={false}
    onClose={[Function]}
    showClose={true}
    title="Edit metadata"
    width="60%"
  >
    <div />
    <div
      className="app-container"
    />
    <Form
      id="edit_metadata"
      isHorizontal={true}
    >
      <FormGroup
        fieldId="product-name"
        isRequired={true}
        label="Product Name"
      >
        <InputGroup>
          <FormSelect
            aria-label="FormSelect Product"
            className=""
            isDisabled={false}
            isValid={true}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            value=""
          >
            <FormSelectOption
              isDisabled={false}
              key="0"
              label="Select a Product"
              value=""
            />
          </FormSelect>
          <FormSelect
            aria-label="FormSelect Version"
            className=""
            id="productVersion"
            isDisabled={false}
            isValid={true}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            value=""
          >
            <FormSelectOption
              isDisabled={false}
              key=""
              label="Select a Version"
              required={false}
              value=""
            />
          </FormSelect>
        </InputGroup>
      </FormGroup>
      <FormGroup
        fieldId="document-usecase"
        isRequired={true}
        label="Document use case"
      >
        <FormSelect
          aria-label="FormSelect Usecase"
          className=""
          isDisabled={false}
          isValid={true}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          value=""
        >
          <FormSelectOption
            isDisabled={false}
            key="0"
            label="Select Use Case"
            value=""
          />
          <FormSelectOption
            isDisabled={false}
            key="1"
            label="Administer"
            value="Administer"
          />
          <FormSelectOption
            isDisabled={false}
            key="2"
            label="Deploy"
            value="Deploy"
          />
          <FormSelectOption
            isDisabled={false}
            key="3"
            label="Develop"
            value="Develop"
          />
          <FormSelectOption
            isDisabled={false}
            key="4"
            label="Install"
            value="Install"
          />
          <FormSelectOption
            isDisabled={false}
            key="5"
            label="Migrate"
            value="Migrate"
          />
          <FormSelectOption
            isDisabled={false}
            key="6"
            label="Monitor"
            value="Monitor"
          />
          <FormSelectOption
            isDisabled={false}
            key="7"
            label="Network"
            value="Network"
          />
          <FormSelectOption
            isDisabled={false}
            key="8"
            label="Plan"
            value="Plan"
          />
          <FormSelectOption
            isDisabled={false}
            key="9"
            label="Provision"
            value="Provision"
          />
          <FormSelectOption
            isDisabled={false}
            key="10"
            label="Release"
            value="Release"
          />
          <FormSelectOption
            isDisabled={false}
            key="11"
            label="Troubleshoot"
            value="Troubleshoot"
          />
          <FormSelectOption
            isDisabled={false}
            key="12"
            label="Optimize"
            value="Optimize"
          />
        </FormSelect>
      </FormGroup>
      <FormGroup
        fieldId="url-fragment"
        isRequired={true}
        label="Vanity URL fragment"
      >
        <InputGroup>
          <InputGroupText
            aria-label="/"
            id="slash"
          >
            <span>
              /
            </span>
          </InputGroupText>
          <TextInput
            aria-label={null}
            className=""
            id="url-fragment"
            isDisabled={false}
            isReadOnly={false}
            isRequired={true}
            isValid={true}
            onChange={[Function]}
            placeholder="Enter URL"
            type="text"
            value=""
          />
        </InputGroup>
      </FormGroup>
      <div>
        <input
          name="productVersion@TypeHint"
          type="hidden"
          value="Reference"
        />
      </div>
    </Form>
  </Modal>
</Fragment>
`;

exports[`Versions tests test changePublishState function 1`] = `undefined`;

exports[`Versions tests test dismissNotification function 1`] = `undefined`;

exports[`Versions tests test fetchProductVersionDetails function 1`] = `Array []`;

exports[`Versions tests test fetchVersions function 1`] = `undefined`;

exports[`Versions tests test getHarrayChildNamed function 1`] = `""`;

exports[`Versions tests test getModuleUrl function 1`] = `"/content/modules.query.json?nodeType=pant:module&where=[urlFragment]=\\"undefined\\""`;

exports[`Versions tests test handleURLInput function 1`] = `undefined`;

exports[`Versions tests test hideSuccessAlert function 1`] = `undefined`;

exports[`Versions tests test loginRedirect function 1`] = `""`;

exports[`Versions tests test onArchiveSelect function 1`] = `undefined`;

exports[`Versions tests test onArchiveToggle function 1`] = `undefined`;

exports[`Versions tests test onChangeProduct function 1`] = `undefined`;

exports[`Versions tests test onChangeUsecase function 1`] = `undefined`;

exports[`Versions tests test onExpandableToggle function 1`] = `undefined`;

exports[`Versions tests test onHeadingToggle function 1`] = `undefined`;

exports[`Versions tests test previewDoc function 1`] = `undefined`;

exports[`Versions tests test renderRedirect function 1`] = `""`;

exports[`Versions tests test saveMetadata function 1`] = `undefined`;
